### PR TITLE
Modernize ranges::for_each

### PR DESCRIPTION
**Pull request by miscco**
This modernizes `ranges::for_each` and `ranges::for_each_n` to the current state of the art

### CODE REVIEW

File: tests/std/tests/P0896R4_ranges_alg_for_each/test.cpp [hunk 1]
```diff
@@ -2,74 +2,63 @@
...
-    STATIC_ASSERT(same_as<decltype(for_each(borrowed<true>{}, identity{})), for_each_result<int*, identity>>);
+struct instantiator {
+    static constexpr P expected[3] = {{1, 42}, {3, 42}, {5, 42}};
 
-    R const values = {{{0, 42}, {2, 42}, {4, 42}}};
-    auto incr      = [](auto& y) { ++y; };
+    template <ranges::input_range ReadWrite>
+    static constexpr void call() {
+        using ranges::for_each, ranges::for_each_result, ranges::in_fun_result, ranges::iterator_t;
+        auto incr = [](auto& y) { ++y; };
```
Comment (CaseyCarter) [inline tests/std/tests/P0896R4_ranges_alg_for_each/test.cpp:30]: Could we please move this lambda into class or namespace scope to reduce total template instantiations?

File: tests/std/tests/P0896R4_ranges_alg_for_each/test.cpp [hunk 2]
```diff
@@ -2,74 +2,63 @@
...
-        same_as<decltype(for_each(borrowed<false>{}, identity{})), for_each_result<ranges::dangling, identity>>);
-    STATIC_ASSERT(same_as<decltype(for_each(borrowed<true>{}, identity{})), for_each_result<int*, identity>>);
+struct instantiator {
+    static constexpr P expected[3] = {{1, 42}, {3, 42}, {5, 42}};
 
-    R const values = {{{0, 42}, {2, 42}, {4, 42}}};
-    auto incr      = [](auto& y) { ++y; };
+    template <ranges::input_range ReadWrite>
+    static constexpr void call() {
+        using ranges::for_each, ranges::for_each_result, ranges::in_fun_result, ranges::iterator_t;
```
Comment (CaseyCarter) [inline tests/std/tests/P0896R4_ranges_alg_for_each/test.cpp:29]: I think `in_fun_result` is unused.

File: tests/std/tests/P0896R4_ranges_alg_for_each_n/test.cpp [hunk 3]
```diff
@@ -2,53 +2,42 @@
...
-    static void call(In in = {}) {
-        using std::iter_difference_t;
+    static constexpr P expected[3] = {{1, 42}, {3, 42}, {5, 42}};
 
-        using Fun = void (*)(std::iter_common_reference_t<In>);
-        ranges::for_each_n(std::move(in), iter_difference_t<In>{}, Fun{});
+    template <indirectly_writable<P> ReadWrite>
+    static constexpr void call() {
+        using ranges::for_each_n, ranges::for_each_n_result, ranges::iterator_t, ranges::distance;
+        auto incr = [](auto& y) { ++y; };
```
Comment (CaseyCarter) [inline tests/std/tests/P0896R4_ranges_alg_for_each_n/test.cpp:23]: Ditto lambda.

File: tests/std/tests/P0896R4_ranges_alg_for_each/test.cpp (hunk 1 above)
Comment (miscco) [reply tests/std/tests/P0896R4_ranges_alg_for_each/test.cpp:30]: Done

File: tests/std/tests/P0896R4_ranges_alg_for_each/test.cpp (hunk 2 above)
Comment (miscco) [reply tests/std/tests/P0896R4_ranges_alg_for_each/test.cpp:29]: It is

File: tests/std/tests/P0896R4_ranges_alg_for_each_n/test.cpp (hunk 3 above)
Comment (miscco) [reply tests/std/tests/P0896R4_ranges_alg_for_each_n/test.cpp:23]: Done

**Comment by miscco** [comment]
Urgh I think adding `constexpr` changes the result type of the lambda.

**Comment by CaseyCarter** [comment]
> Urgh I think adding `constexpr` changes the type of the lambda.

Yes: `constexpr` makes objects implicitly `const`; throwing a `remove_const_t</**/>` around `decltype(lambda)` should fix it.

File: tests/std/tests/P0896R4_ranges_alg_for_each/test.cpp [hunk 4]
```diff
@@ -2,74 +2,66 @@
...
-        ranges::for_each(in, Fun{});
-        ranges::for_each(ranges::begin(in), ranges::end(in), Fun{});
+            auto result = for_each(wrapped_input.begin(), wrapped_input.end(), incr, get_first);
+            STATIC_ASSERT(
+                same_as<decltype(result), for_each_result<iterator_t<ReadWrite>, remove_const_t<decltype(incr)>>>);
+            assert(result.in == wrapped_input.end());
+            assert(ranges::equal(expected, input));
+
+            int some_value = 1729;
+            result.fun(some_value);
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_for_each/test.cpp:44]: I observe that this test verifies that `result.fun` is invocable, but not that it actually contains the state accumulated by being invoked on each element. (An incorrect implementation could return a copy of the original callable object.) No change requested.

File: tests/std/tests/P0896R4_ranges_alg_for_each_n/test.cpp [hunk 5]
```diff
@@ -2,53 +2,42 @@
...
-        ranges::for_each_n(std::move(in), iter_difference_t<In>{}, Fun{});
-
-        using ProjFun = void (*)(unique_tag<0>);
-        ranges::for_each_n(std::move(in), iter_difference_t<In>{}, ProjFun{}, ProjectionFor<In>{});
+    static constexpr P expected[3] = {{1, 42}, {3, 42}, {5, 42}};
+
+    template <indirectly_writable<P> ReadWrite>
+    static constexpr void call() {
+        using ranges::for_each_n, ranges::for_each_n_result, ranges::iterator_t, ranges::distance;
+        { // Validate iterator + sentinel overload
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_for_each_n/test.cpp:25]: Technically, `for_each_n` takes an iterator and a count, not a sentinel.

Comment (CaseyCarter) [reply tests/std/tests/P0896R4_ranges_alg_for_each_n/test.cpp:25]:
```suggestion
        { // Validate iterator overload
```

Comment (miscco) [reply tests/std/tests/P0896R4_ranges_alg_for_each_n/test.cpp:25]: Technically we do not need the whole ting at all as there is no other overload.

**Comment by CaseyCarter** [comment]
Thanks for modernizing each of these algorithms.

**Pull request merged**